include/nbl/builtin/hlsl/bxdf/common.hlsl

```diff
@@ -48,358 +139,976 @@ struct Basic
+        {
+            // TODO: can we optimize?
+            retval.VdotH = nbl::hlsl::dot<vector3_type>(V,H);
+            retval.LdotH = nbl::hlsl::dot<vector3_type>(L,H);
+            retval.NdotH2 = retval.NdotH * retval.NdotH;
+            return true;
+        }
+        return false;
+    }
+    template<class ObserverRayDirInfo, class IncomingRayDirInfo NBL_FUNC_REQUIRES(ray_dir_info::Basic<ObserverRayDirInfo> && ray_dir_info::Basic<IncomingRayDirInfo>)
+    static bool compute(
+        NBL_REF_ARG(this_t) retval,
+        NBL_CONST_REF_ARG(surface_interactions::SIsotropic<ObserverRayDirInfo>) interaction,
+        NBL_CONST_REF_ARG(SLightSample<IncomingRayDirInfo>) _sample,
+        const scalar_type eta, NBL_REF_ARG(vector3_type) H
+    )
+    {
+        const scalar_type NdotV = interaction.getNdotV();
+        const scalar_type NdotL = _sample.getNdotL();
+        const bool transmitted = isTransmissionPath(NdotV,NdotL);
+
+        fresnel::OrientedEtas<scalar_type> orientedEta = fresnel::OrientedEtas<scalar_type>::create(NdotV, eta);
+
+        const vector3_type V = interaction.getV().getDirection();
+        const vector3_type L = _sample.getL().getDirection();
+        const scalar_type VdotL = nbl::hlsl::dot<vector3_type>(V, L);
+        return compute(retval,transmitted,V,L,interaction.getN(),NdotL,VdotL,orientedEta.value,orientedEta.rcp,H);
+    }
+    template<class ObserverRayDirInfo, class IncomingRayDirInfo NBL_FUNC_REQUIRES(ray_dir_info::Basic<ObserverRayDirInfo> && ray_dir_info::Basic<IncomingRayDirInfo>)
+    static bool compute(
+        NBL_REF_ARG(this_t) retval,
+        NBL_CONST_REF_ARG(surface_interactions::SIsotropic<ObserverRayDirInfo>) interaction,
+        NBL_CONST_REF_ARG(SLightSample<IncomingRayDirInfo>) _sample,
+        const scalar_type eta
+    )
+    {
+        vector3_type dummy;
+        return compute<ObserverRayDirInfo, IncomingRayDirInfo>(retval,interaction,_sample,eta,dummy);
+    }
+
+    bool isValidVNDFMicrofacet(const bool is_bsdf, const bool transmission, const scalar_type VdotL, const scalar_type eta, const scalar_type rcp_eta)
+    {
+        return NdotH >= 0.0 && !(is_bsdf && transmission && (VdotL > -min(eta, rcp_eta)));
+    }
+
+    scalar_type getVdotH() NBL_CONST_MEMBER_FUNC { return VdotH; }
+    scalar_type getLdotH() NBL_CONST_MEMBER_FUNC { return LdotH; }
+    scalar_type getNdotH() NBL_CONST_MEMBER_FUNC { return NdotH; }
+    scalar_type getNdotH2() NBL_CONST_MEMBER_FUNC { return NdotH2; }
+
+    scalar_type VdotH;
+    scalar_type LdotH;
+    scalar_type NdotH;
+    scalar_type NdotH2;
 };
 
 
+#define NBL_CONCEPT_NAME AnisotropicMicrofacetCache
+#define NBL_CONCEPT_TPLT_PRM_KINDS (typename)
+#define NBL_CONCEPT_TPLT_PRM_NAMES (T)
+#define NBL_CONCEPT_PARAM_0 (cache, T)
+#define NBL_CONCEPT_PARAM_1 (aniso, typename T::anisotropic_type)
+#define NBL_CONCEPT_PARAM_2 (pNdotL, typename T::scalar_type)
+#define NBL_CONCEPT_PARAM_3 (_sample, typename T::sample_type)
```

Comment:
I just noticed it, but none of the Cache, etc. classes should have typedefs that default to `SDefaultConcept`, so they shouldn't be in the concepts

if you need them to check concept methods, use `SDefaultConcept<..>` instead of `typename T::concept_type` for the `NBL_CONCEPT_PARAM_N`

This applies to every concept after `SLightSample`